org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -207,7 +208,7 @@ public void underlyingFileDeleted(IFileBuffer buffer) {
 	}
 
 	private static class LanguageServerContext {
-		boolean cancelled = false;
+		final AtomicBoolean cancelled = new AtomicBoolean(false);
```

Comment:
wouldn't it have been sufficient to make the boolean volatile plus removing the `synchronized(contextToStop) {` guard?